src/go/util/httppattern/match_sequence_generator.go

```diff
@@ -0,0 +1,193 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package httppattern
+
+import (
+	"fmt"
+	"strings"
+)
+
+// MatchSequenceGenerator generate the match sequence of added methods based on their
+// http pattern. The sequence should be same as the matching order of
+// https://github.com/GoogleCloudPlatform/esp-v2/blob/641ce1d5c177401e424f2b27dd45de1bf797530b/src/api_proxy/path_matcher/path_matcher.h#L1
+type MatchSequenceGenerator struct {
```

Comment:
I suggest we can call it PathMatcher  or PathMetricTrie  or PathMetricTree

Reply:
One thing I don't like about current approach is: we are building so big that most of features are not needed.  http_method is not needed in the tree, the whole idea of the tree is to detect regex pattern overlap and generate proper order for these overlapped url_template.

Definitely, we don't want to check in and maintain dead code (not used code).  "lookup" is dead code.   http_method is not needed in the structure,  we only need to sort url_template. Ideally, only need to sort these "overlapped" ones,  don't care about the ones that not overlapped.

Reply:
can we do this: to simplify this tree
* only put url_template,  not http_method.  not need to store variable binding, not need to deduplicate. 
* this tree is only used for sorting,  not for lookup.  only unit-test the final sorting result.
Based on this url_template order to generate RouteMatch,  
just be care, each url_template may have multiple http_rules, since each http_rule has http_match + url

Reply:
ack

Reply:
I agree this can be simplified, but we need `http_method` in the tree, right? Because we support wildcard HTTP methods, it must be used as a key and matched against.

Here is an example path config:
1. GET /a/**/c
2. *      /a/b/c

The request `GET a/b/c` should match path 1. The request `POST /a/b/c` should match path 2.

If the method is not stored in the tree, how would we determine this?

Reply:
Yes, urlTemplate deduplication + appending wildcard method in the end

Right now they are done in the lib and I guess Wayne wants to move them out of this library. It should be straightforward to be done by scanning once.

Reply:
After checking the code, I prefer to keep the current implementation.

Deduplication out of this library still requires the segments produced by it(serialize it somehow and deduplicate in map). Plus, we also need to handle the `wildcard` method order. The current approach of sorting httpMethod + urlTemplate is relatively clean.

Reply:
I am fine. I can see the benefits of keeping the code in this shape;  it is directly ported from c++,  it should be less buggy than writing something new.